catalog/views.py

```diff
@@ -1,15 +1,52 @@
 from django.contrib import messages
-from django.contrib.auth.mixins import LoginRequiredMixin
-from django.shortcuts import get_object_or_404, render
+from django.contrib.auth.mixins import (LoginRequiredMixin,
+                                        PermissionRequiredMixin,
+                                        UserPassesTestMixin)
+from django.shortcuts import get_object_or_404, redirect, render
 from django.urls import reverse_lazy
-from django.views import View
 from django.views.generic import (CreateView, DeleteView, DetailView, ListView,
-                                  TemplateView, UpdateView)
+                                  TemplateView, UpdateView, View)
 
 from .forms import ProductForm
 from .models import Category, ContactInfo, Product
 
 
+class ProductUnpublishView(LoginRequiredMixin, PermissionRequiredMixin, View):
+    """
+    Представление для снятия продукта с публикации.
+    Доступно только авторизованным пользователям с соответствующим правом.
+    """
+
+    permission_required = "catalog.can_unpublish_product"
+
+    def post(self, request, pk):
+        product = get_object_or_404(Product, pk=pk)
+        product.is_published = False
+        product.save()
```

Comment:
Так чуть оптимальнее - обновляет только одно поле. Иначе в UPDATE передаются все поля модели
```suggestion
        product.save(update_fields=['is_published'])
```